### PR TITLE
Some api docs

### PR DESCRIPTION
**Pull request by fitzgen**

### CODE REVIEW

File: crates/wasmtime/src/store/context.rs [hunk 1]
```diff
@@ -94,6 +94,49 @@ pub trait AsContext {
...
+///     // Therefore, this use of `cx` is a use-after-move and prohibited by the
+///     // borrow checker.
+///     let other_export = instance.get_export(cx, "other_export");
+/// #   drop((my_export, other_export));
+/// }
+/// ```
+///
+/// To fix this, manually insert reborrows like `&mut *cx` that would otherwise
+/// normally be inserted automatically by the Rust compiler for non-generic
+/// methods:
```
Comment (abrown) [inline crates/wasmtime/src/store/context.rs:125]: Thanks for documenting this--super helpful!

File: crates/wasmtime/src/instance.rs [hunk 2]
```diff
@@ -299,6 +299,23 @@ impl Instance {
...
+    /// lazily populated. While `wasmtime` could use interior mutability to
+    /// paper over this, since getting an export is "logically" a non-mutating
+    /// operation, this would bring with it a new problem. `wasmtime` would have
+    /// to choose whether to use a mutex or `RefCell` for the interior
+    /// mutability; the former implies unnecessasry overhead for single-threaded
+    /// usage, while the latter would prohibit `Store` from implementing `Send`
+    /// and `Sync`, making multi-threaded usage impossible.
+    ///
+    /// Given these trade offs -- and because `wasmtime` will not stop lazily
+    /// populating exports, because this makes instantiation faster -- we
```
Comment (sunfishcode) [inline crates/wasmtime/src/instance.rs:315]: Style suggestion: Instead of an absolute negative, "wasmtime will not [...]", say "Lazily populating exports makes instantiation faster, and that's a really good thing for many use cases".

File: crates/wasmtime/src/instance.rs [hunk 3]
```diff
@@ -299,6 +299,23 @@ impl Instance {
...
+    /// decided that the best option is to avoid interior mutability and expose
+    /// what's happening under the covers to users via requiring a mutable
+    /// context.
```
Comment (alexcrichton) [inline crates/wasmtime/src/instance.rs:318]: Personally I find this style of documentation to be helpful for a pretty small subset of people. Most of this can be understood by Rust & Wasmtime experts, but that's not necessarily who's always reading the documentation. I would instead perhaps recommend a style where we explain why the mutable context is used here and some benefits of that, but I wouldn't go into many details of "if we did this other thing it would have these implications".

Basically I don't want to draw too too much attention to this design decision because it's relatively minor compared to the rest of Wasmtime, and while it's important to be aware of it it's not necessarily too useful I think to consider what alternate implementation strategies were rejected when thinking about this. (tbh I never even considered using a lock or a cell, the mutability here just seemed natural)

Comment (fitzgen) [reply crates/wasmtime/src/instance.rs:318]: > (tbh I never even considered using a lock or a cell, the mutability here just seemed natural)

I would argue that this is only because you understand that it is doing lazy population and mutation under the covers though. I think most people would expect that getting an export is just reading a field or doing a hash map lookup.

Otherwise good points.

**Comment by fitzgen** [comment]
Reworded the `get_export` + mutable context docs to be both more "user facing" and not have an absolute negative.

**Review by alexcrichton — APPROVED** [review]
